bugbug/similarity.py

```diff
@@ -268,134 +267,95 @@ def __init__(self, cut_off=0.2, cleanup_urls=True, nltk_tokenizer=False):
 class Word2VecWmdSimilarity(Word2VecSimilarityBase):
     def __init__(self, cut_off=0.2, cleanup_urls=True, nltk_tokenizer=False):
         super().__init__(cleanup_urls=cleanup_urls, nltk_tokenizer=nltk_tokenizer)
-
-    # word2vec.wmdistance calculates only the euclidean distance. To get the cosine distance,
-    # we're using the function with a few subtle changes. We compute the cosine distances
-    # in the get_similar_bugs method and use this inside the wmdistance method.
-    def wmdistance(self, document1, document2, all_distances, distance_metric="cosine"):
-        model = self.w2vmodel
-        if len(document1) == 0 or len(document2) == 0:
-            print(
-                "At least one of the documents had no words that were in the vocabulary. Aborting (returning inf)."
-            )
-            return float("inf")
-
-        dictionary = gensim.corpora.Dictionary(documents=[document1, document2])
-        vocab_len = len(dictionary)
-
-        # Sets for faster look-up.
-        docset1 = set(document1)
-        docset2 = set(document2)
-
-        distance_matrix = np.zeros((vocab_len, vocab_len), dtype=np.double)
-
-        for i, t1 in dictionary.items():
-            for j, t2 in dictionary.items():
-                if t1 not in docset1 or t2 not in docset2:
-                    continue
-
-                if distance_metric == "euclidean":
-                    distance_matrix[i, j] = np.sqrt(
-                        np.sum((model.wv[t1] - model.wv[t2]) ** 2)
-                    )
-                elif distance_metric == "cosine":
-                    distance_matrix[i, j] = all_distances[model.wv.vocab[t2].index, i]
-
-        if np.sum(distance_matrix) == 0.0:
-            print("The distance matrix is all zeros. Aborting (returning inf).")
-            return float("inf")
-
-        def nbow(document):
-            d = np.zeros(vocab_len, dtype=np.double)
-            nbow = dictionary.doc2bow(document)
-            doc_len = len(document)
-            for idx, freq in nbow:
-                d[idx] = freq / float(doc_len)
-            return d
-
-        d1 = nbow(document1)
-        d2 = nbow(document2)
-
-        return emd(d1, d2, distance_matrix)
-
-    def calculate_all_distances(self, words):
-        return np.array(
-            1.0
-            - np.dot(
-                self.w2vmodel.wv.vectors_norm,
-                self.w2vmodel.wv.vectors_norm[
-                    [self.w2vmodel.wv.vocab[word].index for word in words]
-                ].transpose(),
-            ),
-            dtype=np.double,
-        )
+        self.dictionary = Dictionary(self.corpus)
+        self.tfidf = TfidfModel(dictionary=self.dictionary)
 
     def get_similar_bugs(self, query):
```

Comment:
It's a pity that the code became so complex :(
I was hoping by switching to r-wmd library we could simplify it, but it seems like we ended up making it even more complex than before :D
Anyway, there is an improvement in the metrics, so I guess it's a good improvement nevertheless.